### ISSUE TITLE
[Routing] UrlGenerator should try locale xx_XX and xx for i18n routes 

### ISSUE THREAD

**Issue by ampaze** [symfony/symfony#28733]
**Symfony version(s) affected**: 4.1.5

**Description**  
When defining i18n routes, with only a language and not a combination of language and region, the fallback mechanism via _canonical_route does not work, when the locale contains both language and region.

**How to reproduce** 
```yaml
locale: en_US
```
```yaml
petition_show:
    path:
        de: /petitionen/{id}/{slug}
        en: /petitions/{id}/{slug}
```

```php
$this->generateUrl('petition_show', ...
```

Gives the error:
Unable to generate a URL for the named route "petition_show" as such route does not exist.

**Possible Solution**  

UrlGenerator::generate should check both the full locale and only the language part of the locale.
This would have to be reflected in PhpGeneratorDumper::generateGenerateMethod.

**Additional context**  

This would be the same behaviour as the for example the translator works.

**Comment by nicolas-grekas** [comment]
That's already implemented in #27957, or do you mean something else?

**Comment by ampaze** [comment]
> That's already implemented in #27957, or do you mean something else?

You are correct, didn't expect this to be >= Symfony 4.2 and I didn't look in pull requests.

Thanks.
